### PR TITLE
PF372 Prestations souhaitées/préconisées/retenues

### PR DESCRIPTION
**Pull request by qcattez**
Permet de définir si une prestation est souhaitée/préconisée/retenue

Tâche after_party : rake after_party:migrate_prestation_choices

choices

### CODE REVIEW

File: app/views/projets/proposition.html.slim [hunk 1]
```diff
@@ -72,23 +72,35 @@
...
             table.pp-table border="0" cellpadding="0" cellspacing="0" width="100%"
               tbody
                 tr
-                  th.empty scope="col"  &nbsp;
-                  th.top-tab scope="col"  Proposition
-                - @prestations.each do |prestation|
+                  th.empty scope="col"    &nbsp;
+                  th.top-tab scope="col"  Souhaitée
+                  th.top-tab scope="col"  Préconiséee
+                  th.top-tab scope="col"  Sélectionnée
```
Comment (kemenaran) [inline app/views/projets/proposition.html.slim]: Cette colonne s'appelle "Sélectionnée" ici, mais "Retenue" dans la vue synthétique. Il vaudrait mieux harmoniser, non ?

File: db/migrate/20170427135722_remove_unused_columns_from_prestations.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class RemoveUnusedColumnsFromPrestations < ActiveRecord::Migration
+  def change
+    remove_column :prestations, :entreprise
+    remove_column :prestations, :recevable
+    remove_column :prestations, :scenario
+    remove_column :prestations, :souhaite
+    remove_column :prestations, :preconise
+    remove_column :prestations, :retenu
+  end
+end
```
Comment (kemenaran) [inline db/migrate/20170427135722_remove_unused_columns_from_prestations.rb]: 👌

File: db/migrate/20170428095412_add_adjectives_to_prestations_projets.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+class AddAdjectivesToPrestationsProjets < ActiveRecord::Migration
+  def change
+    add_column :prestations_projets, :wished,      :boolean, null: false, default: false
```
Comment (kemenaran) [inline db/migrate/20170428095412_add_adjectives_to_prestations_projets.rb]: En regardant des dictionnaires, j'ai l'impression que la traduction de _"souhaité"_ serait quand même plutôt _"desired"_ ou _"requested"_.

Mais je ne sais pas si c'est important ; je te laisse juge :)

File: lib/tasks/deployment/20170504092142_migrate_prestation_choices.rake [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+      if prestation_choice.projet_id.present? && prestation_choice.prestation_id.present?
+        prestation_choice.update(selected: true) unless prestation_choice.wished || prestation_choice.recommended || prestation_choice.selected
+      end
+    end
+
+    # Update task as completed.  If you remove the line below, the task will
+    # run with every deploy (or every time you call after_party:run).
+    AfterParty::TaskRecord.create version: '20170504092142'
+  end
+end
```
Comment (kemenaran) [inline lib/tasks/deployment/20170504092142_migrate_prestation_choices.rake]: 👌

File: spec/factories/prestation_choices.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+FactoryGirl.define do
+  factory :prestation_choice do
+    association :projet
+    association :prestation
+
+    trait(:wished)      { wished      true }
+    trait(:recommended) { recommended true }
+    trait(:selected)    { selected    true }
+  end
+end
```
Comment (kemenaran) [inline spec/factories/prestation_choices.rb]: 👌

File: spec/features/3_proposition_travaux/remplir_proposition_travaux_spec.rb [hunk 6]
```diff
@@ -193,36 +202,39 @@
...
+        before { projet.prestation_choices << create(:prestation_choice, :selected, projet: projet, prestation: old_used_prestation) }
 
         scenario "j'ai toujours accès à cette prestation" do
           visit dossier_proposition_path(projet)
-          expect(page).not_to have_content('Ancienne prestation non utilisée')
-          expect(page).to have_content('Ancienne prestation utilisée')
-          expect(find_field('Ancienne prestation utilisée')).to be_checked
+          expect(page).not_to have_content old_unused_prestation.libelle
+          expect(page).to     have_content old_used_prestation.libelle
+          expect(find_field("prestation_#{old_used_prestation.id}_selected")).to be_checked
```
Comment (kemenaran) [inline spec/features/3_proposition_travaux/remplir_proposition_travaux_spec.rb]: Avec le recul, il me semble que le test des prestations dépréciées est trop un edge-case pour avoir sa place dans les tests d'acceptance : ça serait vraiment plus un test de controller.

Si tu te sens le courage de le migrer en test de controlleur, je ne dis pas non – mais sinon, comme de toute façon c'était déjà là avant cette PR, on peut aussi le laisser en l'état.

File: spec/lib/tasks/deployment/migrate_prestation_choices_spec.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      expect(prestation_choice_without_prestation.recommended).to be false
+      expect(prestation_choice_without_prestation.selected).to    be false
+
+      prestation_choice_without_projet.reload
+      expect(prestation_choice_without_projet.wished).to      be false
+      expect(prestation_choice_without_projet.recommended).to be false
+      expect(prestation_choice_without_projet.selected).to    be false
+    end
+  end
+end
```
Comment (kemenaran) [inline spec/lib/tasks/deployment/migrate_prestation_choices_spec.rb]: C'est classe, le test de tâche rake 👌 

Je ne sais pas si ça te fait pareil, mais moi ça me donne vachement confiance dans la moulinette : c'est testé, exécuté automatiquement, pas d'intervention manuelle, et tout ☺️

File: spec/lib/tasks/deployment/migrate_prestation_choices_spec.rb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+      prestation_choice.reload
+      expect(prestation_choice.wished).to      be true
+      expect(prestation_choice.recommended).to be false
+      expect(prestation_choice.selected).to    be false
+    end
+  end
+
+  context "quand la table de jointure est erronée" do
+    let!(:prestation_choice_without_prestation) { create :prestation_choice, projet: projet, prestation: nil }
+    let!(:prestation_choice_without_projet)     { create :prestation_choice, projet: nil,    prestation: prestation }
```
Comment (kemenaran) [inline spec/lib/tasks/deployment/migrate_prestation_choices_spec.rb]: Pour moi l'utilisation de nombreux `let!` (avec le point d'exclamation) est plutôt un _code-smell_ – mais dans le cas présent je ne vois pas comment faire mieux, c'est vrai que ça semble nécessaire.

File: app/views/projets/proposition.html.slim (hunk 1 above)
Comment (qcattez) [inline app/views/projets/proposition.html.slim]: Merci t'as raison !

File: spec/lib/tasks/deployment/migrate_prestation_choices_spec.rb (hunk 8 above)
Comment (qcattez) [inline spec/lib/tasks/deployment/migrate_prestation_choices_spec.rb]: A part les mettre dans le `it` ?

File: spec/lib/tasks/deployment/migrate_prestation_choices_spec.rb (hunk 7 above)
Comment (qcattez) [inline spec/lib/tasks/deployment/migrate_prestation_choices_spec.rb]: Ouais pareil pour moi :) Avec after_party en plus c'est beau :D

File: spec/features/3_proposition_travaux/remplir_proposition_travaux_spec.rb (hunk 6 above)
Comment (qcattez) [inline spec/features/3_proposition_travaux/remplir_proposition_travaux_spec.rb]: J'ai regardé et je pense que c'est obligé d'être dans les tests d'acceptance car c'est purement de l'affichage dans la vue et je ne vois pas comment tester ça dans le contrôleur :/

**Comment by qcattez** [comment]
@kemenaran j'ai fais une rebase interactif pour corriger le `wished` en `desired`. Ce fut fastidieux mais c'est passé \o/
Maintenant, il peu y avoir quelques renommage dans des commits qui semblent ne rien avoir à y faire :/

File: app/controllers/dossiers_controller.rb [hunk 9]
```diff
@@ -46,10 +77,80 @@ def recommander_operateurs
...
+      end
+    end
+  end
+
   def export_filename
     "dossiers_#{Time.now.strftime('%Y-%m-%d_%H-%M')}.csv"
   end
 
+  def prestations_with_choices
+    # simplify by .leftjoin when Rails 5
```
Comment (kemenaran) [inline app/controllers/dossiers_controller.rb]:
```
# This query be simplified by using `left_joins` once we'll be running on Rails 5
```

File: spec/lib/tasks/deployment/migrate_prestation_choices_spec.rb (hunk 8 above)
Comment (kemenaran) [inline spec/lib/tasks/deployment/migrate_prestation_choices_spec.rb]: Oui, voilà… La différence de lisibilité ne serait pas flagrante, en même temps. Mais le mettre dans le `it`, ça permet aux gens qui ne connaissent pas l'usage un peu finaud de `let!` de comprendre le test quand même.

Je dirais qu'on laisse comme ça pour l'instant, et qu'on se repose la question la prochaine fois qu'on en aura besoin :)

**Comment by kemenaran** [comment]
@qcattez bon pour moi 👍 

Je te laisse merger toi-même quand tu auras pu éditer le commentaire sur Rails 5 ?
